mllib/src/main/scala/org/apache/spark/mllib/classification/LogisticRegression.scala

```diff
@@ -61,20 +67,70 @@ class LogisticRegressionModel (
 
   override protected def predictPoint(dataMatrix: Vector, weightMatrix: Vector,
       intercept: Double) = {
-    val margin = weightMatrix.toBreeze.dot(dataMatrix.toBreeze) + intercept
-    val score = 1.0 / (1.0 + math.exp(-margin))
-    threshold match {
-      case Some(t) => if (score > t) 1.0 else 0.0
-      case None => score
+    // If dataMatrix and weightMatrix have the same dimension, it's binary logistic regression.
+    if (dataMatrix.size == weightMatrix.size) {
+      val margin = dot(weights, dataMatrix) + intercept
+      val score = 1.0 / (1.0 + math.exp(-margin))
+      threshold match {
+        case Some(t) => if (score > t) 1.0 else 0.0
+        case None => score
+      }
+    } else {
+      val dataWithBiasSize = weightMatrix.size / (nClasses - 1)
+      val dataWithBias = if(dataWithBiasSize == dataMatrix.size) {
+        dataMatrix
+      }  else {
+        assert(dataMatrix.size + 1 == dataWithBiasSize)
+        MLUtils.appendBias(dataMatrix)
+      }
+
+      val margins = Array.ofDim[Double](nClasses)
+
+      val weightsArray = weights match {
+      case dv: DenseVector => dv.values
+      case _ =>
+        throw new IllegalArgumentException(
+          s"weights only supports dense vector but got type ${weights.getClass}.")
+      }
+
+      var i = 0
+      while (i < nClasses - 1) {
+        var margin = 0.0
+        dataWithBias.foreachActive { (index, value) =>
+          if (value != 0.0) margin += value * weightsArray((i * dataWithBiasSize) + index)
+        }
+        margins(i + 1) = margin
+        i += 1
+      }
+
+      /**
+       * Find the one with maximum margins. Note that `margins(0) == 0`.
+       *
+       * PS, if you want to compute the probabilities for each outcome instead of the outcome
+       * with maximum probability, remember to subtract the maxMargin from margins if maxMargin
+       * is positive to prevent overflow.
+       */
+      var label = 0.0
+      var max = margins(0)
```

Comment:
Can this whole stanza be `margins.indexOf(margins.max)`? Granted that needs two passes, but this is a small collection.